File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch07/Python/P07-MatInv.py
# Check of matrix inversion using LU decomposition
from random import *
from linsys import *
from matutil import *

n = 5                                                       # order of matrix
a = [[0]*(n+1) for i in range(n+1)]             # original matrix and inverse
b = [[0]*(n+1) for i in range(n+1)]                      # backup of original
c = [[0]*(n+1) for i in range(n+1)]                            # check matrix

for i in range(1,n+1):                               # generate random matrix
   for j in range(1,n+1): a[i][j] = random()

print("Original matrix:")
MatPrint(a,n,n)

MatCopy(a,b,n,n)                                     # backup original matrix

det = MatInv(a,n)                                           # invert original
if (det == 0e0): print("Singular matrix"); exit(1)

print("Inverse matrix:")
MatPrint(a,n,n)

print("Check A^(-1)A = I:")
MatProd(a,b,c,n,n,n)                         # multiply inverse with original
MatPrint(c,n,n)                                          # print check matrix

File: Introduction_to_numerical_programming_using_Python_and_CPP_Beu/Ch07/C/P07-MatInv.cpp
// Check of matrix inversion using LU decomposition
#include <stdio.h>
#include "memalloc.h"
#include "linsys.h"
#include "matutil.h"

int main()
{
   double **a, **b, **c, det;
   int i, j, n;

   n = 5;                                                  // order of matrix
   a = Matrix(1,n,1,n);                        // original matrix and inverse
   b = Matrix(1,n,1,n);                                 // backup of original
   c = Matrix(1,n,1,n);                                       // check matrix

   for (i=1; i<=n; i++)                             // generate random matrix
      for (j=1; j<=n; j++) a[i][j] = rand();

   printf("Original matrix:\n");
   MatPrint(a,n,n);

   MatCopy(a,b,n,n);                                // backup original matrix

   MatInv(a,n,det);                                        // invert original
   if (det == 0e0) { printf("Singular matrix\n"); exit(1); }

   printf("Inverse matrix:\n");
   MatPrint(a,n,n);

   printf("Check A^(-1)A = I:\n");
   MatProd(a,b,c,n,n,n);                    // multiply inverse with original
   MatPrint(c,n,n);                                     // print check matrix
}
